controller/test/pid_lib/pid_test.cpp

```diff
@@ -0,0 +1,393 @@
+  float Kp = 5.5;
+  float Ki = 1.1;
+  float Kd = 1.5;
+  double output = 0;
+  PID myPID(&input, &output, &setpoint, Kp, Ki, Kd, DIRECT);
+  myPID.SetSampleTime(sample_time);
+  myPID.SetMode(AUTOMATIC);
+  myPID.Compute();
+
+  double last_output = output;
+  // check that call with no time change has no effect
+  myPID.Compute();
+  EXPECT_EQ(output, last_output);
+
+  Time last_time = Hal.now();
+
+  // chose dt to not be a divisor of SampleTime, for better coverage
+  double dt = 6;
+  // Run PID a few times with fast and check output doesn't change unless a
+  // sample time has passed
+  for (int i = 0; i < 100; i++) {
+    Hal.delay(dt);
+    myPID.Compute();
+    if (Hal.now() >= last_time + milliseconds(sample_time)) {
+      last_output = output;
+      last_time = last_time + milliseconds(sample_time);
+    } else {
+      EXPECT_EQ(output, last_output);
+    }
+  }
+}
+
+TEST(pidTest, TaskJitter) {
+  // This test uses integral to check the effect of time between calls on the
+  // PID output. Introducing jitter in call frequency and checking that the
+  // integral takes this jitter into account.
+  float Kp = 0;
+  float Ki = 0.5;
+  float Kd = 0;
+  double output = 0;
+  PID myPID(&input, &output, &setpoint, Kp, Ki, Kd, DIRECT);
+  myPID.SetSampleTime(sample_time);
+  myPID.SetMode(AUTOMATIC);
+  myPID.Compute();
+  // First task, with only Ki set, output = error*sample_time
+  double integral = (setpoint - input) * (sample_time / 1000.0);
+  EXPECT_NEAR(output, integral * Ki, OUTPUT_TOLERANCE);
+  // Advance time with jitter
+  uint32_t dt = sample_time + max_task_jitter;
+  Hal.delay(dt);
+  myPID.Compute();
+  // Expect output to take jitter into account in integral
+  integral += (setpoint - input) * (dt / 1000.0);
+  EXPECT_NEAR(output, integral * Ki, OUTPUT_TOLERANCE);
+  // Advance time and compensate previous jitter to have total time
+  // elapsed = 2*sample time
+  dt = sample_time - max_task_jitter;
+  integral += (setpoint - input) * (dt / 1000.0);
+  Hal.delay(dt);
+  myPID.Compute();
+  EXPECT_NEAR(output, integral * Ki, OUTPUT_TOLERANCE);
+}
+
+TEST(pidTest, SampleTimeChange) {
+  // This test uses integral to check the effect of changing sample time
+  // during execution of the PID
+  float Kp = 0;
+  float Ki = 1.1;
+  float Kd = 0;
+  double output = 0;
+  PID myPID(&input, &output, &setpoint, Kp, Ki, Kd, DIRECT);
+  myPID.SetSampleTime(sample_time);
+  myPID.SetMode(AUTOMATIC);
+  myPID.Compute();
+
+  // First task, with only Ki set, output = error*sample_time
+  double integral = (setpoint - input) * (sample_time / 1000.0);
+  EXPECT_NEAR(output, integral * Ki, OUTPUT_TOLERANCE);
+  // Advance time
+  uint32_t dt = sample_time;
+  Hal.delay(dt);
+  myPID.Compute();
+  // Expect output to integrate this
+  integral += (setpoint - input) * (dt / 1000.0);
+  EXPECT_NEAR(output, integral * Ki, OUTPUT_TOLERANCE);
+  // Change sample time and advance
+  sample_time = 50;
+  dt = sample_time;
```

Comment:
In this test, dt and sample_time always hold identical values, can we get rid of one of them?

Reply:
sure... this is actually a remnant of my (failed) attempt at unifying all those tests.